fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -1971,6 +2121,216 @@ struct ShardedRocksDBKeyValueStore : IKeyValueStore {
+			TraceEvent(SevInfo, "ShardedRocksDBRestoreBegin", logId)
+			    .detail("Path", a.path)
+			    .detail("Checkpoints", describe(a.checkpoints));
+
+			ASSERT(!a.checkpoints.empty());
+
+			PhysicalShard* ps = nullptr;
+			for (const KeyRange& range : a.ranges) {
+				ps = a.shardManager->addRange(range, a.shardId);
+			}
+
+			const CheckpointFormat format = a.checkpoints[0].getFormat();
+			for (int i = 1; i < a.checkpoints.size(); ++i) {
+				if (a.checkpoints[i].getFormat() != format) {
+					throw invalid_checkpoint_format();
+				}
+			}
+
+			rocksdb::Status status;
+			rocksdb::WriteBatch writeBatch;
+			if (format == DataMoveRocksCF) {
+				CheckpointMetaData& checkpoint = a.checkpoints.front();
+				std::sort(a.ranges.begin(), a.ranges.end(), RangeLessThan());
+				std::sort(checkpoint.ranges.begin(), checkpoint.ranges.end(), RangeLessThan());
+				if (a.ranges.empty() || checkpoint.ranges.empty() || a.ranges.size() > checkpoint.ranges.size() ||
+				    a.ranges.front().begin != checkpoint.ranges.front().begin) {
+					TraceEvent(SevError, "ShardedRocksDBRestoreFailed", logId)
+					    .detail("Path", a.path)
+					    .detail("Ranges", describe(a.ranges))
+					    .detail("Checkpoints", describe(a.checkpoints));
+					a.done.sendError(failed_to_restore_checkpoint());
+					return;
+				}
+
+				for (int i = 0; i < a.ranges.size(); ++i) {
+					if (a.ranges[i] != checkpoint.ranges[i] && i != a.ranges.size() - 1) {
+						TraceEvent(SevError, "ShardedRocksDBRestoreFailed", logId)
+						    .detail("Path", a.path)
+						    .detail("Ranges", describe(a.ranges))
+						    .detail("Checkpoints", describe(a.checkpoints));
+						a.done.sendError(failed_to_restore_checkpoint());
+						return;
+					}
+				}
+
+				ASSERT(!ps->initialized());
+				TraceEvent(SevDebug, "ShardedRocksRestoringCF", logId)
+				    .detail("Path", a.path)
+				    .detail("Checkpoints", describe(a.checkpoints))
+				    .detail("RocksDBCF", getRocksCF(a.checkpoints[0]).toString());
+
+				status = ps->restore(a.checkpoints[0]);
+
+				if (!status.ok()) {
+					logRocksDBError(status, "Restore");
+					a.done.sendError(statusToError(status));
+					return;
+				} else {
+					ASSERT(ps->initialized());
+					(*columnFamilyMap)[ps->cf->GetID()] = ps->cf;
+					TraceEvent(SevInfo, "RocksDBRestoreCFSuccess", logId)
+					    .detail("Path", a.path)
+					    .detail("ColumnFaminly", ps->cf->GetName())
+					    .detail("Checkpoints", describe(a.checkpoints));
+
+					// Remove the extra data.
+					KeyRangeRef cRange(a.ranges.back().end, checkpoint.ranges.back().end);
+					if (!cRange.empty()) {
+						TraceEvent(SevInfo, "RocksDBRestoreCFRemoveExtraRange", logId)
+						    .detail("Path", a.path)
+						    .detail("Checkpoints", describe(a.checkpoints))
+						    .detail("RestoreRanges", describe(a.ranges))
+						    .detail("ClearExtraRange", cRange);
+						writeBatch.DeleteRange(ps->cf, toSlice(cRange.begin), toSlice(cRange.end));
+					}
```

Comment:
Are we assuming the imported CF could contain extra data? Should we clear all the gaps between the requested ranges?

Reply:
The actual data ranges are defined in `CheckpointMetaData::ranges`, while the ranges we intend to restore is defined by `RestoreAction::ranges`, so we only need to clear the extra range(s) in `CheckpointMetaData::ranges`